content/blog/2016/2016-08-29-sauce-pipeline.adoc

```diff
@@ -0,0 +1,271 @@
+
+image::/images/post-images/2016-08-29/platforms-report.png[JUnit Report Showing Added Platforms, role="center"]
+
+=== Conclusion: To Awesome and Beyond
+Considering the complexity of the system, I was impressed with how easy it was to integrate Jenkins with Sauce OnDemand to start testing on multiple browsers.
+The plugin worked flawlessly with Jenkins Pipeline.
+I went ahead and ran some additional tests to show that failure reporting also behaved as expected.
+
+[source,groovy]
+----
+//----8<----8<----8<----8<----8<----8<----
+    sh "./node_modules/.bin/nightwatch -e ${configs}" // <1>
+//----8<----8<----8<----8<----8<----8<----
+----
+<1> Removed `--test` filter to run all tests
+
+[NOTE]
+--
+You can view full code for this blog post in the
+link:https://github.com/bitwiseman/JS-Nightwatch.js/tree/sauce-pipeline[sauce-pipeline]
+branch of the
+link:https://github.com/bitwiseman/JS-Nightwatch.js[bitwiseman/JS-Nightwatch.js]
+repository.
+--
+
+image::/images/post-images/2016-08-29/tests-report.png[Tests, Platorms, Results, role="center"]
+
+=== Epilogue: Pipeline vs. Freestyle
+Just for comparison here's the final state of this job in Freestyle UI versus fully-commented pipeline code:
+
+[NOTE]
+--
+This includes the
+link:https://wiki.jenkins-ci.org/display/JENKINS/AnsiColor+Plugin[AnsiColor Plugin]
+to support Nightwatch.js' default ANSI color output.
+--
+
+==== Freestyle
+image::/images/post-images/2016-08-29/freestyle-1.png[Freestyle Job Configuration - SCM, role="center"]
+
+image::/images/post-images/2016-08-29/freestyle-2.png[Freestyle Job Configuration - Wrappers and Sauce, role="center"]
+
+image::/images/post-images/2016-08-29/freestyle-3.png[Freestyle Job Configuration - Build and Publish, role="center"]
+
+
+==== Pipeline
+
+[source,groovy]
+----
+node {
+    stage "Build"
+    checkout scm
+
+    // Install dependencies
+    sh 'npm install'
+
+    stage "Test"
+
+    // Add sauce credentials
+    sauce('f0a6b8ad-ce30-4cba-bf9a-95afbc470a8a') {
+        // Start sauce connect
+        sauceconnect(options: '', useGeneratedTunnelIdentifier: false, verboseLogging: false) {
+
+            // List of browser configs we'll be testing against.
+            def platform_configs = [
+                'chrome',
+                'firefox',
+                'ie',
+                'edge'
+            ].join(',')
+
+            // Nightwatch.js supports color ouput, so wrap this step for ansi color
+            wrap([$class: 'AnsiColorBuildWrapper', 'colorMapName': 'XTerm']) {
+
+                // Run selenium tests using Nightwatch.js
+                // Ignore error codes. The junit publisher will cover setting build status.
+                sh "./node_modules/.bin/nightwatch -e ${platform_configs} || true"
+            }
+
+            junit 'reports/**'
+
+            step([$class: 'SauceOnDemandTestPublisher'])
+        }
+    }
+}
+----
```

Comment:
I think it's worth adding a statement at the end, reiterating why pipeline as code is useful in comparison, allowing the `Jenkinsfile` to be checked in and iterated upon to make it easy to modify and expand your browser test coverage.